keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+
+### Risks and Mitigations
+
+The new validation should increase the security of Kubernetes by
+making it impossible to have ambiguously-interpretable IPs/CIDRs in
+the future.
+
+The most obvious risk to users is that by tightening validation, we
+might break existing clusters. Especially, the fact that we are
+enforcing tighter validation for new objects of existing types means
+that we might break some users' existing workflows / automation that
+were generating now-invalid values. This could be mitigated by having
+API warnings for a few releases before we flip the switch, or by
+having a `Deprecated` feature gate that users could disable if needed
+while they update their infrastructure.
+
+The new validation logic (to allow legacy invalid values) will be more
+complicated than the old logic, and thus potentially more likely to
+have bugs.
+
+## Design Details
+
+Assuming all of the UNRESOLVED sections are ignored, this is already
+implemented, in [PR #122550]; the KEP is being written retroactively
+to make sure we have agreement on that plan (or not).
+
+[PR #122550]: https://github.com/kubernetes/kubernetes/pull/122550
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[ ] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+We will add validation for the new IP/CIDR validation functions, and
+make sure that all of the updated fields have unit tests, and that we
+validate the behavior of being able to make unrelated changes to
+objects with pre-existing invalid values, and that we validate the
+ability to fix invalid fields of immutable objects.
+
+- `k8s.io/apimachinery/pkg/util/validation`: `2024-10-02` - `91.9`
+- `k8s.io/kubernetes/pkgs/apis/core/validation`: `2024-10-02` - `84.3`
+- `k8s.io/kubernetes/pkgs/apis/discovery/validation`: `2024-10-02` - `98.8`
+- `k8s.io/kubernetes/pkgs/apis/networking/validation`: `2024-10-02` - `91.4`
+
+##### Integration tests
+
+No new tests, and we will remove
+`test/integration/apiserver/cve_2021_29923_test.go`, which tests that
+it is possible to create new objects with invalid IP values (since it
+won't be possible any more).
+
+##### e2e tests
+
+No new tests, and we will remove `test/e2e/network/funny_ips.go`,
```

Comment:
but if I use cilium or a Service implementation that is not kube-proxy, how do I test that I don't break those legacy Services?

Reply:
You'll need your own unit tests I guess?

I don't really like losing the e2e test, but I wasn't sure what else we could do unless we added some sort of terrible "don't validate the service fully if it has this annotation" hack.

Reply:
we can add one of this Feature tags things, so we skip it in our CI but keep available for others projects to consume, having the code there will not hurt

Reply:
No, the problem is that in order to test what the service proxy does with invalid IPs, you need to feed it a Service/EndpointSlice that contains invalid IPs. But the apiserver won't let us create one now...